File: python/report_writer.py
from electronic_sensor import ElectronicSensor
from sensor_utils import calculate_efficiency


def write_report(filename, assets):
    lines = []
    lines.append("INDUSTRIAL ASSET REPORT")
    lines.append("=======================\n")

    for asset in assets:
        lines.append(f"Name          : {asset.name}")
        lines.append(f"Serial Number : {asset.serial_number}")
        lines.append(f"Base Value    : ${asset.base_value:.2f}")

        if isinstance(asset, ElectronicSensor):
            eff = calculate_efficiency(asset.current_reading, 100.0)
            lines.append(f"Voltage       : {asset.voltage:.1f} V")
            lines.append(f"Reading       : {asset.current_reading:.2f}")
            lines.append(f"Efficiency    : {eff:.1f}%")

        lines.append("")

    content = "\n".join(lines)

    # Write to file
    try:
        with open(filename, "w") as f:
            f.write(content)
        print(f"Report saved to: {filename}")
    except IOError as e:
        print(f"[ERROR] Could not write report: {e}")

File: java/ReportWriter.java
import java.io.FileWriter;
import java.io.IOException;
import java.util.List;

public class ReportWriter {

    public static void writeReport(String filename, List<IndustrialAsset> assets) {
        StringBuilder sb = new StringBuilder();
        sb.append("INDUSTRIAL ASSET REPORT\n");
        sb.append("=======================\n\n");

        for (IndustrialAsset asset : assets) {
            sb.append("Name          : ").append(asset.getName()).append("\n");
            sb.append("Serial Number : ").append(asset.getSerialNumber()).append("\n");
            sb.append(String.format("Base Value    : $%.2f%n", asset.getBaseValue()));

            if (asset instanceof ElectronicSensor) {
                ElectronicSensor s = (ElectronicSensor) asset;
                double eff = SensorUtils.calculateEfficiency(s.getCurrentReading(), 100.0);
                sb.append(String.format("Voltage       : %.1f V%n", s.getVoltage()));
                sb.append(String.format("Reading       : %.2f%n", s.getCurrentReading()));
                sb.append(String.format("Efficiency    : %.1f%%%n", eff));
            }
            sb.append("\n");
        }

        // Write to file
        try (FileWriter fw = new FileWriter(filename)) {
            fw.write(sb.toString());
            System.out.println("Report saved to: " + filename);
        } catch (IOException e) {
            System.out.println("[ERROR] Could not write report: " + e.getMessage());
        }
    }
}
